users/helauth/pipeline.py

```diff
@@ -102,3 +102,66 @@ def update_user_from_profile(request: WSGIRequest, *, user: User | None = None)
         user.date_of_birth = birthday_info["birthday"]
 
     user.save()
+
+
+def migrate_user_from_tunnistamo_to_keycloak(
+    backend: TunnistamoOIDCAuth,
+    request: WSGIRequest,
+    response: OIDCResponse,
+    user: User | None = None,
+    **kwargs: Unpack[ExtraKwargs],
+) -> dict[str, Any]:
+    if user is None:
+        return {"user": user}
+
+    id_token = IDToken.from_string(response["id_token"])
+    if (
+        id_token is not None  # There is an id token.
+        and id_token.iss.endswith("helsinki-tunnistus")  # Issuer is keycloak.
+        and id_token.is_ad_login  # It is an AD login.
+        and user.email not in ("", None, "ANON.ANONYMIZED@anonymized.net")  # User has a valid email.
+    ):
+        migrate_from_tunnistamo_to_keycloak(email=user.email)
+
+    return {"user": user}
+
+
+@SentryLogger.log_if_raises("Keycloak migration: Failed to migrate user from tunnistamo to keycloak")
+def migrate_from_tunnistamo_to_keycloak(*, email: str) -> None:
+    # Look at the two most recent AD users with the same email address (no profile ID).
+    # There can be more, but the most recent one is the new Keycloak user and the other one
+    # is the old Tunnistamo user after the auth-migration (when auth moved to backend), which we want to migrate.
+    users: list[User] = list(User.objects.filter(email=email, profile_id="").order_by("date_joined")[:2])
+    if len(users) < 2:
+        return
+
+    old_user = users[0]
+    new_user = users[1]
+
+    # Don't run migration more than once.
+    if not old_user.is_active:
```

Comment:
I don't believe this is necessary. Duplicate AD users should be from the time of the previous migration (3 clients to centralized to backend), in which case the penultimate user (more recent `date_joined`) is from after the migration (the latest user being the new Keycloack user). This user is only created if the user has logged in since the migration. Older users probably aren't using the system anymore.

Reply:
That is generally but not entirely uniformly true, as there are some duplicate AD users with the same email address:
- (A) both identities have been created _subsequent_ to the centralization -- in fact the newest such incarnation appears to have been created last week for an AD user who had previously logged in around March 2024 (with another identity) -- however, examination proves that they are not an issue, as the user incarnation with more recent `date_joined` value is clearly valid;
- (B) or both identities have been created _prior_ to that date (this is the rarer case), and it turns out that for few of them the more recent identity is now a bogus one. (This can be found out in more ways than one.)

**TL;DR**: the present check `if not old_user.is_active` should nevertheless be sufficient, together with the filtering out of non-empty profile ids further above. I went through the three dozen or so such user pairs in production environment (same hel.fi email, both still flagged active, profile id empty in both) and removed the active flag from one of them.

**Longer version**:
Tunnistamo token doesn't contain AD username but the difference in AD groups seems to indicate that for set **A**, person's role is now different and maybe these are cases where an external AD account has been recreated; although it doesn't really matter what the ultimate reason is.
Of set **B**, most cases (6/9) were explained by the more recent  `date_joined` (thus with greater `user.id`) having been used strictly for Django-admin login before the centralization, but few (3/9) could not be readily explained since the other version had never logged in subsequent to that previous migration. It can be speculated that either there had been a Tunnistamo malfunction and it had given out different sub claim / mis-identity for a time (this kind of thing has happened) or suomifi-login had been performed by the user once (by mistake). Whatever the situation, the bogus versions are no longer flagged as active, and will be skipped accordingly.

Reply:
If there are any cases where there are currently two users, and the most recently created one is not the one that should be migrated, then this won't work for those users, since we aren't filtering out inactive users. Currently this just takes the two most recent users, and if the most recent one is inactive, then we just don't perform the migration even once. The new user doesn't get their permissions from the actual user, which is was the one before the inactive one. Did I understand correctly that there were some like that?

Reply:
With the `is_active` check now in the user query this should be fine.
In practical terms existing production environment duplicates of users with an `hel.fi` email have been set as inactive in all cases as revealed by the following SQL (if either one already was not):
`select u1.email, u2.id as newer_id, u1.id as older_id, u2.is_active as newer_active, u1.is_active as older_active, u2.username as newer_username, u1.username as older_username, u2.date_joined as newer_joined, u2.last_login as newer_last_login, u1.date_joined as older_joined, u1.last_login as older_last_login from public.user u1, public.user u2 where u1.id < u2.id and u1.email = u2.email and u1.email like '%hel.fi' and u1.profile_id = '' and u2.profile_id = '' order by u2.id desc, u1.id desc;`
In most cases the earlier version of the user (which dated to time before the authentication centralization) is in any case inactive; but in few cases it is the newer user that has been set as inactive, either because it had only been used for Django login before that date or was some other bogus variant which had never logged since that date.

**So**: if the migration query now still finds more than two active users ('new', 'old', plus 'older'), it would be because someone has continued using the system or otherwise has different sub-claim from Tunnistamo in later September (possible, there's one such case since yesterday) -- and that should be okay, since in those cases `order by date_joined desc` can be assumed to be fetch the two valid ones.